source/Halibut/Transport/TcpConnectionFactory.cs

```diff
@@ -47,6 +43,27 @@ public IConnection EstablishNewConnection(ExchangeProtocolBuilder exchangeProtoc
 
             return new SecureConnection(client, ssl, exchangeProtocolBuilder, log);
         }
+        
+        public async Task<IConnection> EstablishNewConnectionAsync(ExchangeProtocolBuilder exchangeProtocolBuilder, ServiceEndPoint serviceEndpoint, ILog log, CancellationToken cancellationToken)
+        {
+            log.Write(EventType.OpeningNewConnection, $"Opening a new connection to {serviceEndpoint.BaseUri}");
+
+            var certificateValidator = new ClientCertificateValidator(serviceEndpoint);
+            var client = await CreateConnectedTcpClientAsync(serviceEndpoint, log, cancellationToken);
+            log.Write(EventType.Diagnostic, $"Connection established to {client.Client.RemoteEndPoint} for {serviceEndpoint.BaseUri}");
+
+            var stream = client.GetStream();
+
+            log.Write(EventType.SecurityNegotiation, "Performing TLS handshake");
+            var ssl = new SslStream(stream, false, certificateValidator.Validate, UserCertificateSelectionCallback);
+            await ssl.AuthenticateAsClientAsync(serviceEndpoint.BaseUri.Host, new X509Certificate2Collection(clientCertificate), SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12, false);
```

Comment:
No cancellation support?

Reply:
Not with net48, yes with net6 although the method signature is different.

Reply:
Should we "patch" net48 so we can support cancellation across both?

Reply:
Did we have a standard way of patching these methods to look like they support cancellation.

Reply:
I have currently added a `// TODO - ASYNC ME UP!` comment here. To ensure we don't forget about this.

Reply:
> Did we have a standard way of patching these methods to look like they support cancellation

Would this be a use case of the 2 tasks, the action task and a cancellation/timeout task, if the cancellation/timeout task wins, then close the stream?

Reply:
WaitAsync doesn't exist either in net48